### PR TITLE
Use ranking query in custom dimensions archiver

### PR DESCRIPTION
**Pull request by diosmosis**
### Description:

As title. There might be a random test failure since RankingQuery doesn't apply ordering (probably a good thing).

### Review

* [ ] Functional review done
* [ ] Potential edge cases thought about (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] Usability review done (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] Security review done [see checklist](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] Code review done
* [ ] Tests were added if useful/possible
* [ ] Reviewed for breaking changes
* [ ] Developer changelog updated if needed
* [ ] Documentation added if needed
* [ ] Existing documentation updated if needed

### CODE REVIEW

File: core/API/DocumentationGenerator.php [hunk 1]
```diff
@@ -314,6 +314,10 @@ public function getExampleUrl($class, $methodName, $parametersToSet = array())
             } // if there isn't a default value for a given parameter,
             // we need a 'know default value' or we can't generate the link
             elseif ($defaultValue instanceof NoDefaultValue) {
+                if (defined('PIWIK_TEST_MODE') && $nameVariable != 'idSubtable') {
+                    // print error message since it is annoying to have to debug to find this
+                    print "$nameVariable required in $class::$methodName but no value given (this usually means your API test config is missing some parameters)\n";
```
Comment (sgiehl) [inline core/API/DocumentationGenerator.php]: That is now printed quite often within various working tests.

File: core/DataAccess/LogAggregator.php [hunk 2]
```diff
@@ -1059,9 +1065,11 @@ protected function getActionsMetricFields()
      * @param bool|string $where An optional SQL expression used in the SQL's **WHERE** clause.
      * @param array $additionalSelects Additional SELECT fields that are not included in the group by
      *                                 clause. These can be aggregate expressions, eg, `SUM(somecol)`.
+     * @param RankingQuery|bool $rankingQuery
      * @return \Zend_Db_Statement
      */
-    public function queryConversionsByDimension($dimensions = array(), $where = false, $additionalSelects = array(), $extraFrom = [])
+    public function queryConversionsByDimension($dimensions = array(), $where = false, $additionalSelects = array(), $extraFrom = [],
+                                                $rankingQuery = false, $rankingQueryGenerate = false)
```
Comment (sgiehl) [inline core/DataAccess/LogAggregator.php:1076]: The last parameter isn't documented. And from looking at the code it wasn't directly visible what a difference it makes. But seems when the ranking query is executed it directly returns all records where as otherwise it returns a query pointer, right?

File: plugins/CustomDimensions/Archiver.php [hunk 3]
```diff
@@ -195,6 +217,19 @@ public function queryCustomDimensionActions(DataArray $dataArray, $valueField, $
...
         $logAggregator = $this->getLogAggregator();
         $query     = $logAggregator->generateQuery($select, $from, $where, $groupBy, $orderBy);
+
+        if ($this->rankingQueryLimit > 0) {
+            $rankingQuery = new RankingQuery($this->rankingQueryLimit);
+            $rankingQuery->addLabelColumn(array($valueField, 'url'));
+            // TODO: combine below 4
+            $rankingQuery->addColumn(Metrics::INDEX_PAGE_SUM_TIME_SPENT, 'sum');
+            $rankingQuery->addColumn(Metrics::INDEX_BOUNCE_COUNT, 'sum');
+            $rankingQuery->addColumn(Metrics::INDEX_PAGE_EXIT_NB_VISITS, 'sum');
```
Comment (sgiehl) [inline plugins/CustomDimensions/Archiver.php]: Guess those line can be removed, as they are actually contained in `$metricIds`?

File: core/DataAccess/LogAggregator.php [hunk 4]
```diff
@@ -421,7 +422,7 @@ public function generateQuery($select, $from, $where, $groupBy, $orderBy, $limit
         return $query;
     }
 
-    protected function getVisitsMetricFields()
+    public function getVisitsMetricFields()
```
Comment (sgiehl) [inline core/DataAccess/LogAggregator.php]: Any reason for making this public? Couldn't find any additionally usage that requires it 🤔

**Review by sgiehl — CHANGES_REQUESTED** [review]
There are also some tests failing that don't look random

File: core/DataAccess/LogAggregator.php (hunk 2 above)
Comment (diosmosis) [reply core/DataAccess/LogAggregator.php:1076]: Yes that is correct, I'll add docs.

File: core/DataAccess/LogAggregator.php (hunk 4 above)
Comment (diosmosis) [reply core/DataAccess/LogAggregator.php]: I was using it, forgot to change it back :+1:

File: core/API/DocumentationGenerator.php (hunk 1 above)
Comment (diosmosis) [reply core/API/DocumentationGenerator.php]: I'll do this in another pr.

File: plugins/CustomDimensions/Archiver.php (hunk 3 above)
Comment (diosmosis) [reply plugins/CustomDimensions/Archiver.php]: :+1:

File: core/DataAccess/LogAggregator.php [hunk 5]
```diff
@@ -1075,6 +1087,17 @@ public function queryConversionsByDimension($dimensions = array(), $where = fals
         $orderBy = false;
         $query   = $this->generateQuery($select, $from, $where, $groupBy, $orderBy);
 
+        if (!empty($rankingQuery)) {
+            $sumColumns = array_keys($availableMetrics);
+            $rankingQuery->addColumn($sumColumns, 'sum');
```
Comment (tsteur) [inline core/DataAccess/LogAggregator.php:1092]: @diosmosis can we assume that it is always a `sum`? or is it maybe also sometimes min, max, etc?

File: plugins/CustomDimensions/Archiver.php [hunk 6]
```diff
@@ -245,4 +276,15 @@ protected function cleanCustomDimensionValue($value)
         return self::LABEL_CUSTOM_VALUE_NOT_DEFINED;
     }
 
+    private function getRankingQueryLimit()
+    {
+        $configGeneral = Config::getInstance()->General;
+        $configLimit = $configGeneral['archiving_ranking_query_row_limit'];
+        $limit = $configLimit == 0 ? 0 : max(
```
Comment (tsteur) [inline plugins/CustomDimensions/Archiver.php:298]: I suppose this looks good. Was wondering if the 50K ranking query limit could cause issues since we technically allow 1000 root tables and 1000 entries per subtable by default. I'm assuming this will limit the number of results shown quite a bit but that's kind of also the point of this ranking query? 

For the ranking query limit would maybe do something like 

```php
$this->rankingQueryLimit = max(configLimit, 10 * $this->maximumRowsInDataTableLevelZero);
```

just in case someone has a high `datatable_archiving_maximum_rows_custom_dimensions` set and so you can still get different results per subtable but I think also not really needed.

I'm not entirely sure what the report groups by though and how the subtables are created so could also just leave as is.

File: core/DataAccess/LogAggregator.php (hunk 5 above)
Comment (diosmosis) [reply core/DataAccess/LogAggregator.php:1092]: missed this :+1:

File: plugins/CustomDimensions/Archiver.php (hunk 6 above)
Comment (diosmosis) [reply plugins/CustomDimensions/Archiver.php:298]: subtables are urls, so it makes sense :+1:
